File: tests/pass219/test_pass219_lane5_executable_capability_self_model_1_43.c
#include "hhs_pass219_lane5_executable_capability_self_model_1_43.h"

#include <stdio.h>
#include <string.h>

#define CHECK(expr) do { \
    if (!(expr)) { \
        fprintf(stderr, "CHECK failed at %s:%d: %s\n", __FILE__, __LINE__, #expr); \
        return 1; \
    } \
} while (0)

static HHSExactPass219Lane5CapabilitySelfModelDescriptorV1 valid_descriptor(void) {
    HHSExactPass219Lane5CapabilitySelfModelDescriptorV1 value;
    memset(&value, 0, sizeof(value));
    value.struct_size = (uint32_t)sizeof(value);
    value.version = HHS_EXACT_PASS219_LANE5_CAPABILITY_SELF_MODEL_VERSION;
    value.namespace_id = HHS_EXACT_PASS219_LANE5_CAPABILITY_SELF_MODEL_NAMESPACE;
    value.total_entries = 3U;
    value.public_entries = 1U;
    value.native_entries = 2U;
    value.restricted_entries = 0U;
    value.canonical_boundary_entries = 1U;
    value.public_catalog_signature64 = UINT64_C(101);
    value.native_export_signature64 = UINT64_C(102);
    value.dependency_signature64 = UINT64_C(103);
    value.model_signature64 = UINT64_C(104);
    value.canonical_boundary_signature64 = UINT64_C(33);
    value.entry_signature64[0] = UINT64_C(11);
    value.entry_signature64[1] = UINT64_C(22);
    value.entry_signature64[2] = UINT64_C(33);
    value.source_kind[0] = HHS_EXACT_PASS219_LANE5_CAPABILITY_SOURCE_PUBLIC_REGISTRY;
    value.source_kind[1] = HHS_EXACT_PASS219_LANE5_CAPABILITY_SOURCE_NATIVE_EXACT_ABI;
    value.source_kind[2] = HHS_EXACT_PASS219_LANE5_CAPABILITY_SOURCE_NATIVE_EXACT_ABI;
    value.authority_class[0] = HHS_EXACT_PASS219_LANE5_CAPABILITY_AUTH_OBSERVATION;
    value.authority_class[1] = HHS_EXACT_PASS219_LANE5_CAPABILITY_AUTH_GOVERNED_TRANSFORM;
    value.authority_class[2] = HHS_EXACT_PASS219_LANE5_CAPABILITY_AUTH_CANONICAL_ADMISSION_BOUNDARY;
    value.public_registry_complete = 1U;
    value.native_exact_abi_complete = 1U;
    value.ordered_unique_identity = 1U;
    value.candidate_only = 1U;
    value.requires_signed_environmental_vm81_admission = 1U;
    return value;
}

static int rejected(HHSExactPass219Lane5CapabilitySelfModelDescriptorV1 value) {
    HHSExactPass219Lane5CapabilitySelfModelReceiptV1 receipt;
    memset(&receipt, 0, sizeof(receipt));
    return hhs_exact_pass219_lane5_capability_self_model_validate(&value, &receipt) != HHS_EXACT_STATUS_OK;
}

int main(void) {
    HHSExactPass219Lane5CapabilitySelfModelAuthorityV1 authority;
    HHSExactPass219Lane5CapabilitySelfModelDescriptorV1 descriptor;
    HHSExactPass219Lane5CapabilitySelfModelReceiptV1 receipt;
    HHSExactPass219Lane5CapabilitySelfModelDescriptorV1 bad;

    memset(&authority, 0, sizeof(authority));
    CHECK(hhs_exact_pass219_lane5_capability_self_model_version() ==
          HHS_EXACT_PASS219_LANE5_CAPABILITY_SELF_MODEL_VERSION);
    CHECK(hhs_exact_pass219_lane5_capability_self_model_authority(&authority) == HHS_EXACT_STATUS_OK);
    CHECK(authority.global_capability_discovery == 1U);
    CHECK(authority.public_registry_snapshot == 1U);
    CHECK(authority.native_exact_abi_snapshot == 1U);
    CHECK(authority.canonical_boundary_singleton == 1U);
    CHECK(authority.candidate_only == 1U);
    CHECK(authority.canonical_vm81_mutation_authority == 0U);
    CHECK(authority.canonical_hash72_authority == 0U);
    CHECK(authority.canonical_hash216_authority == 0U);
    CHECK(authority.canonical_persistence_authority == 0U);
    CHECK(authority.requires_signed_environmental_vm81_admission == 1U);

    descriptor = valid_descriptor();
    memset(&receipt, 0, sizeof(receipt));
    CHECK(hhs_exact_pass219_lane5_capability_self_model_validate(&descriptor, &receipt) == HHS_EXACT_STATUS_OK);
    CHECK(receipt.accepted == 1U);
    CHECK(receipt.total_entries == 3U);
    CHECK(receipt.public_entries == 1U);
    CHECK(receipt.native_entries == 2U);
    CHECK(receipt.canonical_boundary_entries == 1U);
    CHECK(receipt.canonical_boundary_signature64 == UINT64_C(33));
    CHECK(receipt.descriptor_signature64 != 0U);
    CHECK(receipt.receipt_signature64 != 0U);
    CHECK(receipt.candidate_only == 1U);
    CHECK(receipt.canonical_vm81_mutation_authority == 0U);
    CHECK(receipt.canonical_persistence_authority == 0U);
    CHECK(receipt.requires_signed_environmental_vm81_admission == 1U);

    bad = descriptor;
    bad.entry_signature64[1] = 0U;
    CHECK(rejected(bad));

    bad = descriptor;
    bad.entry_signature64[1] = bad.entry_signature64[0];
    CHECK(rejected(bad));

    bad = descriptor;
    bad.entry_signature64[1] = UINT64_C(44);
    CHECK(rejected(bad));

    bad = descriptor;
    bad.source_kind[0] = 99U;
    CHECK(rejected(bad));

    bad = descriptor;
    bad.authority_class[0] = 99U;
    CHECK(rejected(bad));

    bad = descriptor;
    bad.public_entries = 2U;
    CHECK(rejected(bad));

    bad = descriptor;
    bad.model_signature64 = 0U;
    CHECK(rejected(bad));

    bad = descriptor;
    bad.dependency_signature64 = 0U;
    CHECK(rejected(bad));

    bad = descriptor;
    bad.authority_class[2] = HHS_EXACT_PASS219_LANE5_CAPABILITY_AUTH_GOVERNED_TRANSFORM;
    bad.canonical_boundary_entries = 0U;
    CHECK(rejected(bad));

    bad = descriptor;
    bad.authority_class[1] = HHS_EXACT_PASS219_LANE5_CAPABILITY_AUTH_CANONICAL_ADMISSION_BOUNDARY;
    bad.canonical_boundary_entries = 2U;
    CHECK(rejected(bad));

    bad = descriptor;
    bad.canonical_boundary_signature64 = UINT64_C(22);
    CHECK(rejected(bad));

    bad = descriptor;
    bad.canonical_vm81_mutation_authority = 1U;
    CHECK(rejected(bad));

    bad = descriptor;
    bad.canonical_persistence_authority = 1U;
    CHECK(rejected(bad));

    bad = descriptor;
    bad.requires_signed_environmental_vm81_admission = 0U;
    CHECK(rejected(bad));

    printf(
        "PASS219_LANE5_CAPABILITY_SELF_MODEL_PASS total=%u public=%u native=%u descriptor=%llu receipt=%llu\n",
        receipt.total_entries,
        receipt.public_entries,
        receipt.native_entries,
        (unsigned long long)receipt.descriptor_signature64,
        (unsigned long long)receipt.receipt_signature64
    );
    return 0;
}

File: tests/pass219/test_pass219_lane5_executable_capability_self_model_1_43.py
from __future__ import annotations

from pathlib import Path

from hhs_backend.runtime.hhs_pass219_lane5_executable_capability_self_model_1_43 import (
    CANONICAL_BOUNDARY_EXPORT,
    Pass219Lane5ExecutableCapabilitySelfModel,
)
from hhs_python.runtime.hhs_pass219_lane5_capability_self_model_bridge import (
    AUTH_CANONICAL_ADMISSION_BOUNDARY,
    AUTH_RESTRICTED_OR_UNAVAILABLE,
)


def test_live_repository_capability_self_model_is_exact_and_candidate_only() -> None:
    builder = Pass219Lane5ExecutableCapabilitySelfModel()
    model = builder.build()

    assert model["counts"]["public_registry"] > 0
    assert model["counts"]["native_exact_abi"] > 0
    assert model["counts"]["total"] == len(model["nodes"])
    assert model["counts"]["canonical_boundaries"] == 1
    assert model["scope"]["pass147_public_registry_complete"] is True
    assert model["scope"]["cumulative_exact_abi_complete"] is True
    assert model["scope"]["repository_total_historical_capability_complete"] is False

    node_ids = [node["node_id"] for node in model["nodes"]]
    signatures = [node["entry_signature64"] for node in model["nodes"]]
    assert len(node_ids) == len(set(node_ids))
    assert len(signatures) == len(set(signatures))

    boundary = [
        node
        for node in model["nodes"]
        if node["authority_class"] == AUTH_CANONICAL_ADMISSION_BOUNDARY
    ]
    assert len(boundary) == 1
    assert boundary[0]["export_name"] == CANONICAL_BOUNDARY_EXPORT

    receipt = model["native_receipt"]
    assert receipt["accepted"] is True
    assert receipt["total_entries"] == model["counts"]["total"]
    assert receipt["public_entries"] == model["counts"]["public_registry"]
    assert receipt["native_entries"] == model["counts"]["native_exact_abi"]
    assert receipt["canonical_boundary_entries"] == 1
    assert receipt["candidate_only"] is True
    assert receipt["canonical_vm81_mutation_authority"] is False
    assert receipt["canonical_hash72_authority"] is False
    assert receipt["canonical_hash216_authority"] is False
    assert receipt["canonical_persistence_authority"] is False
    assert receipt["requires_signed_environmental_vm81_admission"] is True

    authority = model["authority"]
    assert authority["global_capability_discovery"] == 1
    assert authority["public_registry_snapshot"] == 1
    assert authority["native_exact_abi_snapshot"] == 1
    assert authority["canonical_boundary_singleton"] == 1
    assert authority["candidate_only"] == 1
    assert authority["canonical_vm81_mutation_authority"] == 0
    assert authority["canonical_persistence_authority"] == 0
    assert authority["requires_signed_environmental_vm81_admission"] == 1


def test_capability_self_model_replays_to_identical_roots_and_native_receipt() -> None:
    builder = Pass219Lane5ExecutableCapabilitySelfModel()
    first = builder.build()
    second = builder.build()

    for key in (
        "public_catalog_root_hash72",
        "native_export_root_sha256",
        "dependency_root_sha256",
        "model_root_sha256",
    ):
        assert first[key] == second[key]

    for key in (
        "descriptor_signature64",
        "receipt_signature64",
        "canonical_boundary_signature64",
    ):
        assert first["native_receipt"][key] == second["native_receipt"][key]


def test_transitive_native_export_discovery_is_rooted_at_aggregate_header(tmp_path: Path) -> None:
    include = tmp_path / "hhs_runtime" / "include"
    include.mkdir(parents=True)
    (include / "hhs_runtime_exact_abi.h").write_text(
        '#include "child.h"\nHHS_EXACT_API int hhs_exact_root_probe(void);\n',
        encoding="utf-8",
    )
    (include / "child.h").write_text(
        '#include "grandchild.h"\nHHS_EXACT_API int hhs_exact_child_probe(void);\n',
        encoding="utf-8",
    )
    (include / "grandchild.h").write_text(
        'HHS_EXACT_API int hhs_exact_grandchild_probe(void);\n',
        encoding="utf-8",
    )
    (include / "unreachable.h").write_text(
        'HHS_EXACT_API int hhs_exact_unreachable_probe(void);\n',
        encoding="utf-8",
    )

    builder = Pass219Lane5ExecutableCapabilitySelfModel(tmp_path, bridge=object())  # type: ignore[arg-type]
    exports = builder._native_exports()
    names = [item["export_name"] for item in exports]
    assert names == [
        "hhs_exact_child_probe",
        "hhs_exact_grandchild_probe",
        "hhs_exact_root_probe",
    ]
    assert "hhs_exact_unreachable_probe" not in names


def test_restricted_public_classification_is_preserved_when_present() -> None:
    builder = Pass219Lane5ExecutableCapabilitySelfModel()
    catalog = builder._public_catalog()
    live_nodes = [builder._public_node(item)[0] for item in catalog]
    assert len(live_nodes) == len(catalog)

    synthetic = {
        "capability_id": "PUB-RESTRICTED-PROBE",
        "capability_hash72": "0" * 72,
        "surface_type": "CLI",
        "classification": "EXPLICITLY_RESTRICTED_BY_CONTRACT",
        "capabilities": [],
        "reversibility_class": "REJECTED_AS_UNSAFE",
        "mutating": False,
        "argv": ["restricted-probe"],
        "parameters": [],
        "description": "classification preservation probe",
    }
    node, edges = builder._public_node(synthetic)
    assert node["authority_class"] == AUTH_RESTRICTED_OR_UNAVAILABLE
    assert node["source_kind"] == "PUBLIC_REGISTRY"
    assert edges == []
